Add keyspace storage options

psarna opened the pull request:
This series is part of the shared storage project.

The STORAGE option is designed to hold a map of options
used for customizing storage for given keyspace.
The option is kept in a system_schema.scylla_keyspaces table.

This option is guarded with a schema feature, because it's kept in a new schema table: `system_schema.scylla_keyspaces`.

Example of the contents of the new table:
```cql
cassandra@cqlsh> select * from system_schema.scylla_keyspaces;

 keyspace_name | storage_options                                | storage_type
---------------+------------------------------------------------+--------------
           ksx | {'bucket': '/tmp/xx', 'endpoint': 'localhost'} |           S3
```
Native storage options are not kept in the table, as this format doesn't hold any extra options and it would therefore just be a waste of storage.

raphaelsc commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -96,14 +96,35 @@ std::map<sstring, sstring> ks_prop_defs::get_replication_options() const {
+        set_field(opts.bucket, "bucket");
+        set_field(opts.key_id, "key_id");
+        set_field(opts.endpoint, "endpoint");
+    }
```
let's fail create ks statement if unrecognized option is found

nyh replied:
Also, the three specific items you copy here (bucket, key_id, endpoint) only makes sense for the S3 type, not for the native part. So why don't we just copy all the options here instead of these specific three, and just the CQL statement to set the options will validate that the combination makes sense, i.e, that with S3 you can specify "bucket" but not "doghouse", and with NATIVE you can't specify neither bucket nor doghouse.

Of course later as we add more types and options, we'll need additional cluster options so you can't set a new type (say "NFS") if not all the nodes on the cluster support it. What a mess :-( But at least the mess will be in one place (the CQL statement validation) and not all over the place as in this function?

psarna replied:
Back in the hackathon days, it was decided that these "storage options" should simply hold a union of all options needed for different formats that we support. E.g. for a shared mountpoint we can still use the "bucket" part to indicate a local path (which is logically a bucket), and key_id can also make sense for optional authentication. Native storage simply doesn't use any options, which is fine. And if any new storage driver needs another option, it can be added to this structure.

nyh replied:
If you go with this approach, I think you need to be a bit more forward-looking with the options and their names - e.g., why call something a "bucket" if this name only has meaning for S3? For example, we could call it "location" and then explain in the comment that this is a type-specific string which describes where to store the files - e.g., in s3 it refers to a bucket.

I don't know what key_id was supposed to mean. Is it an authentication key of some sort? There is just one of those (Amazon has two?)?

Anyway, what if I add an NFS storage option and find myself needing a "doghouse" option, but then someone sets type=S3 but doghouse=7? Which part of the code will tell me I did something wrong here?

nyh replied:
By the way, this discussion reminded me of the idea of URL. It was invented exactly to encapsulate these things - a URL http://some.domain.com/some/directory a "type" (e.g., http://), "endpoint" (some.domain.com/) and directory (some/directory) - and also additional parameters and optional port. 
I'm not saying you should use a URL here, but it's a thought.

psarna replied:
> something like https://github.com/scylladb/scylla/blob/master/compaction/compaction_descriptor.hh#L52 should do the work. we have to guarantee that each type performs its own validation on its supported options

yeah, I'm already in the middle of implementing this kind of approach. I'll push something today

raphaelsc commented on db/schema_tables.cc:
```diff
@@ -1745,7 +1787,41 @@ lw_shared_ptr<keyspace_metadata> create_keyspace_from_schema_partition(const sch
+                    };
+                    set_field(storage_opts.bucket, "bucket");
+                    set_field(storage_opts.key_id, "key_id");
+                    set_field(storage_opts.endpoint, "endpoint");
```
we should define constants for storage options, to avoid duplicating them

raphaelsc commented on db/schema_tables.cc:
```diff
@@ -1690,6 +1717,16 @@ std::vector<mutation> make_create_keyspace_mutations(lw_shared_ptr<keyspace_meta
+        if (!storage_map.empty()) {
+            store_map(scylla_m, ckey, "storage", timestamp, storage_map);
+        }
+        mutations.emplace_back(std::move(scylla_m));
```
I asked above to fail create ks statement if unrecognized option is found, but at least we're ignoring them here

nyh commented on db/schema_features.hh:
```diff
@@ -22,14 +22,16 @@ enum class schema_feature {
     COMPUTED_COLUMNS,
     CDC_OPTIONS,
     PER_TABLE_PARTITIONERS,
+    SCYLLA_KEYSPACES,
```
Why do we need the extra complication of a new "Scylla Keyspaces" table? Couldn't we add more columns to the existing "keyspaces" table? 
Is the thinking that in the future Cassandra might add columns to "keyspaces" too and if they accidentally choose the same name as we did, we'll be screwed? In this case won't it be easier to add columns with "scylla_" in their names?
Or maybe the thinking is that some application may read "keyspaces" and be surprised with the extra columns?

nyh replied:
Ok, these reasons make sense, so indeed let's go with your separate-system-table approach.

By the way, I noted in my review that even though it is a separate table, it too would need some sort of forward-looking plan, as in the future you may want to add more storage types and options for the existing types, and need to forbid using new types or options until the entire cluster allows them, and so on - so I think some of your code which duplicates the logic of which options exist and which don't in multiple places and maybe should be written in a way that the verification should be done only in one place (e.g., one verification function called by the CQL statement) and other places in the code could just work with string maps without knowledge of which options exist in many different places in the code.

nyh commented on data_dictionary/storage_options.hh:
```diff
@@ -0,0 +1,47 @@
+    };
+
+    storage_type type;
+    sstring bucket;
```
The options below only makes sense for S3, right? If so, maybe you should use a union or std::variant or something? Or at least, a comment say this (if you don't care about wasting a few bytes)?

psarna replied:
Yes, we could use a std::variant, but it would pretty much waste the same amount of space (actually, at least 1 more byte for storing the variant id). If in the future we'll be supporting more storage formats that need an entirely different set of options, it would make sense to rewrite it to a variant.

nyh commented on db/schema_tables.cc:
```diff
@@ -254,6 +255,32 @@ schema_ptr keyspaces() {
+        // regular column name type
+        utf8_type,
+        // comment
+        "scylla-specific keyspace definitions"
```
This phrasing is a bit ambigous - are those "scylla-specific-keyspace definitions" or "scylla-specific keyspace-definitions"? :-) In other words, do these define Scylla-specific keyspaces, *or*, add Scylla-specific features to keyspaces which already should be listed in the regular "keyspaces" table?

I think the answer is the latter - the same keyspace will need to be listed in both "keyspaces" and "scylla_keyspaces".

psarna commented:
v2:
 - rewrote to std::variant, looks way more elegant and deduplicated now, thanks for the suggestion
 - scylla_keyspaces now holds two separate columns - storage_type and storage_options, since it's easier to distinguish the type from its specific options

Still a draft, for the same reasons as mentioned in the issue description.

psarna commented:
Update: `schema_change_test` **does not** need to be updated, because before the upgrade there won't be any keyspaces that have custom storage options, and storage options would only be accepted if all nodes already support the feature (that part still needs to be done on CQL level, I'll do it in a minute). Combined with the fact that native storage does not generate any entries in scylla_keyspaces schema table, the schema is actually not affected during upgrade.

psarna commented:
rebased; no functional changes pushed

avikivity commented on data_dictionary/data_dictionary.cc:
```diff
@@ -243,6 +261,50 @@ std::vector<view_ptr> keyspace_metadata::views() const {
 }
 
+storage_options::value_type storage_options::from_map(std::string_view type, std::map<sstring, sstring> values) {
+    if (type == "NATIVE") {
```
Let's have a descriptive name, like "LOCAL". Native works if you understand the project history, but coming from a clean slate you wouldn't guess what native means.

Also, let's reject such changes in ALTER (haven't checked if you do already).

avikivity commented on db/schema_tables.cc:
```diff
@@ -3134,6 +3161,9 @@ std::vector<schema_ptr> all_tables(schema_features features) {
     }
+    if (features.contains<schema_feature::SCYLLA_KEYSPACES>()) {
+        result.emplace_back(scylla_keyspaces());
+    }
```
Please check if something needs to be done to get Raft to manage this table when enabled. @gleb-cloudius @kbr-

avikivity replied:
Still missing, no?

psarna replied:
Yes, I guess I'll reach out to one of the Raft people directly

psarna replied:
Kamil told me that as long as I add more mutations and use the existing announcing code, which I'm pretty sure I do, it will be handled by Raft correctly. I also verified by enabling Raft locally and observing that CREATE KEYSPACE and ALTER KEYSPACE statements with storage options specified are correctly recorded in group0_history table.

avikivity replied:
But is the new table managed by raft?

avikivity replied:
Ah, I understand the answer now.

@kbr- @kostja we should adjust DML so that writes into these tables go through the leader and we don't risk inconsistency.

avikivity commented on data_dictionary/storage_options.hh:
```diff
@@ -0,0 +1,47 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
SPDX

avikivity commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -50,6 +51,21 @@ void cql3::statements::alter_keyspace_statement::validate(query_processor& qp, c
+            data_dictionary::storage_options new_options = _attrs->get_storage_options();
+            if (new_options.type_string() != current_options.type_string()) {
+                throw exceptions::invalid_request_exception(format("Cannot alter storage options type from {} to {}",
+                        current_options.type_string(), new_options.type_string()));
```
Ah, you do check it.

There is risk of someone altering the bucket name and losing access, but I can see valid use cases for it.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1772,7 +1791,27 @@ lw_shared_ptr<keyspace_metadata> create_keyspace_from_schema_partition(const sch
+    // Scylla-specific info is fetched below
+    if (features.contains<schema_feature::SCYLLA_KEYSPACES>()) {
+        auto keyspace_key = dht::decorate_key(*scylla_keyspaces(), partition_key::from_singular(*scylla_keyspaces(), keyspace_name));
+        auto&& rs = co_await db::system_keyspace::query(proxy, NAME, SCYLLA_KEYSPACES, keyspace_key);
```
This makes the function non-pure. Can we move it to the caller?

psarna replied:
Technically we could, and that was my initial approach, but I dropped because there are several callers and **not** fetching this info would be a bug, which would be extremely trivial to miss if I suddenly start requiring the caller to do it. It would be extremely easy to miss some of the changes during backports. I gave it some thought and ultimately decided that if I absolutely require something to be called before this function, it's better to just add it to the function body, so that nobody can forget to do it. If you still prefer me to try and split it, let me know.

psarna replied:
update: I'll try to slightly refactor this function's signature to try and get rid of this co_await. I'll let you know if I succeeded or not, and if not I'll just come back here and discuss why.

psarna replied:
done!

avikivity commented:
Where is  the Cql.g change?

docs/?

--experimental?

psarna commented:
> Where is the Cql.g change?

There is none. Keyspace options are already stored in a map, and `STORAGE` simply became another accepted key in this map (ks_prop_defs).
> 
> docs/?

Will do.

> 
> --experimental?

Will do.

avikivity commented:
> Update: `schema_change_test` **does not** need to be updated, because before the upgrade there won't be any keyspaces that have custom storage options, and storage options would only be accepted if all nodes already support the feature (that part still needs to be done on CQL level, I'll do it in a minute). Combined with the fact that native storage does not generate any entries in scylla_keyspaces schema table, the schema is actually not affected during upgrade.

We need to make sure the scylla_keyspaces table is part of the hash.

psarna commented:
> > Update: `schema_change_test` **does not** need to be updated, because before the upgrade there won't be any keyspaces that have custom storage options, and storage options would only be accepted if all nodes already support the feature (that part still needs to be done on CQL level, I'll do it in a minute). Combined with the fact that native storage does not generate any entries in scylla_keyspaces schema table, the schema is actually not affected during upgrade.
> 
> We need to make sure the scylla_keyspaces table is part of the hash.

Indeed, but I think I'll try to craft a separate test case for that, because the support is experimental and I'd rather avoid regenerating all the sstables in the existing schema test based on an experimental feature that can be subject to change.

psarna commented:
v4:
 - renamed `NATIVE` to `LOCAL`
 - added an experimental flag
 - added a docs entry

Still todo: add a test which verifies that `system_schema.scylla_keyspaces` entries are taken into account when computing schema digest

psarna commented:
v5:
 - added a test case for schema digest verification after the new system_schema table is added

nyh commented on data_dictionary/data_dictionary.cc:
```diff
@@ -243,6 +261,50 @@ std::vector<view_ptr> keyspace_metadata::views() const {
+            return "S3";
+        }
+    }, value);
+}
```
I think this implementation is fine, but just want to comment that in the good (?) all days, we used to do this kind of things with inheritance, not std::variant. It allowed a future developer to add a new storage type (e.g., EBS, or whatever) by mostly adding new code in a new file, instead of having to modify a dozen places in an existing file.
But for now, I think what you did is good enough, no need to change.

nyh commented on test/cql-pytest/test_keyspace.py:
```diff
@@ -191,6 +191,59 @@ def creates(count):
+def test_storage_options_s3(cql, scylla_only):
+    ksdef = "WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : '1' } " \
+            "AND STORAGE = { 'type' : 'S3', 'bucket' : '42', 'endpoint' : 'localhost' }"
+    with new_test_keyspace(cql, ksdef) as keyspace:
```
I don't understand something. When S3 **is** implemented, won't this creation of a keyspace try to do something on S3 - such as validating that the bucket exists, or creating an empty index file in the bucket, or something? If there is chance it will do this (is there?), this test will stop working. So isn't it safer to do this test with LOCAL instead of S3?

psarna replied:
That's a valid point, yes. I'll switch to local for now, and later I'll probably go with SHARED_FILESYSTEM, which is yet another backend that's already implemented in one of our hackathon branches, and which uses a local path (which could technically be a mounpoint for a distributed filesystem) for storing all the sstables.

nyh commented on test/cql-pytest/test_keyspace.py:
```diff
@@ -191,6 +191,59 @@ def creates(count):
+    ksdef = "WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : '1' } " \
+            "AND STORAGE = { 'type' : 'S3', 'bucket' : '42', 'endpoint' : 'localhost' }"
+    with new_test_keyspace(cql, ksdef) as keyspace:
+        # It's fine to amend if storage options aren't changed or just the parameters change
```
I find it surprising that it's fine. How can it be fine to switch buckets in an existing table that already contains data? In what scenario will this make sense? If it makes sense to change the bucket, why doesn't it make sense to switch between local and s3?

psarna replied:
True, let's forbid it for now. If we ever find out that it's useful (for, I don't know, migrating to another location, another provider or whatnot), we can allow it back.

psarna commented on data_dictionary/data_dictionary.cc:
```diff
@@ -243,6 +261,50 @@ std::vector<view_ptr> keyspace_metadata::views() const {
+                throw std::runtime_error(format("Unknown storage option for S3: {}", key));
+            }
+        }
+        return options;
```
Right, I think we don't want to allow that. I'll add a test case for missing items and fix the code accordingly.

psarna commented:
v6:
 - disallowed altering storage options, until we figure out the semantics for "changing an S3 bucket"
 - disallowed passing any options for "LOCAL" storage
 - made both bucket and endpoint parameters required for S3
 - changed the tests to not rely on S3 actually working - they may break when we actually add S3 support and validation

nyh commented:
@psarna the series has conflicts and needs rebasing.
I already "approved" this series, but two other reviewers (@avikivity and @raphaelsc) reviewed this series and didn't approve or disapprove it so I don't know if everything was answered to their satisfaction.

psarna commented:
(weekly rebase)
